### DISCUSSION TITLE
Bulk insert with nullable decimals

### DISCUSSION THREAD

**Discussion by megasuperlexa** [DarkWanderer/ClickHouse.Client#547] [Q&A]
I am inserting into a table which has the following column
`balance_after Nullable(Decimal(18, 8))`
However when I am doing bulk insert, the exception is thrown 

> at ClickHouse.Client.Copy.ClickHouseBulkCopy.WriteToServerAsync(IEnumerable1 rows, IReadOnlyCollection1 columns, CancellationToken token)
> 
> at ClickHouse.Client.Types.DecimalType.Write(ExtendedBinaryWriter writer, Object value)
> 
> at System.DBNull.System.IConvertible.ToDecimal(IFormatProvider provider)
> 
> ---> System.InvalidCastException: Object cannot be cast from DBNull to other types.
> 
> ClickHouse.Client.Copy.ClickHouseBulkCopySerializationException: Error when serializing data

which is I guess because `ClickHouse.Client.Types.DecimalType` does not support converting nulls into decimal.
Is this a bug? is there a different type I can enforce for this column in the library? can I somehow trick `ClickHouseBulkCopy` to not assume column types from the database like it does currently but to treat values as `string` for example?

**Comment by megasuperlexa** [comment]
Turns out Clickhouse might return columns in unpredictable order on schema request in `ClickHouseBulkCopy.InitAsync` if column names are not specified beforehand. So when I use `WriteToServerAsync` with `DataTable` overload, the internal column ordinals representation of ClickHouseBulkCopy command does not match that of supplied `DataTable`. This not obvious from the outside and quite dangerous, as I ended up writing data into wrong columns on the server. Hard to notice if most columns have the same datatype like `string`.

**Reply by DarkWanderer** [reply]
Hi. Thanks for reporting this, your concern makes sense. The challenge in amending that behavior may be that there is existing code which already does not check columns, and introducing the remapping may break it. The solution I have in mind is to add a new method which has a specific behavior. Do you think it will solve your case?
